### PR TITLE
Merge pull request #3050 from ChristopherDavenport/cookieJar

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
Future Improvements

- More Complete Test Coverage
- Perhaps an internal counter for the cookiejar to clear all the expired cookies - Currently we clear on every request which is likely overkill.

### CODE REVIEW

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            now <= t._2.expiresAt
+          }
+        )
+      } yield out
+
+    override def evictAll: F[Unit] = ref.set(Map.empty)
+
+    override def addCookie(c: ResponseCookie, uri: Uri): F[Unit] =
+      for {
+        now <- currentHttpDate
```
Comment (rossabaker) [inline client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: We're going to retrieve the date once per cookie?  It's probably cheap enough, but I'd consider a private method into which the date is passed, to support both `addCookie` and `enrichRequest`.

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      } yield out
+
+    override def enrichRequest[N[_]](r: Request[N]): F[Request[N]] =
+      for {
+        cookies <- ref.get.map(_.map(_._2.cookie).toList)
+      } yield cookiesForRequest(r, cookies)
+        .foldLeft(r) { case (req, cookie) => req.addCookie(cookie) }
+  }
+
+  private[middleware] final case class CookieKey(
```
Comment (rossabaker) [inline client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: Is `CookieJar` not tight enough scope here?

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    def apply(
+        setAt: HttpDate,
+        expiresAt: HttpDate,
+        cookie: ResponseCookie
+    ): CookieValue = new CookieValue(setAt, expiresAt, cookie)
+  }
+
+  private[middleware] def currentHttpDate[F[_]: Clock: MonadError[?[_], Throwable]] =
+    Clock[F]
+      .monotonic(SECONDS)
```
Comment (rossabaker) [inline client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: Why not `realTime` here?

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  private[middleware] def responseCookieToRequestCookie(r: ResponseCookie): RequestCookie =
+    RequestCookie(r.name, r.content)
+
+  private[middleware] def cookieAppliesToRequest[N[_]](
+      r: Request[N],
+      c: ResponseCookie): Boolean = {
+
+    val domainApplies = c.domain.exists(s =>
+      r.uri.host.forall { authority =>
+        authority.renderString.contains(s)
```
Comment (rossabaker) [inline client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: This makes me embarrassed of our `Uri`.

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala (hunk 3 above)
Comment (ChristopherDavenport) [reply client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: Whoopsie.

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala (hunk 2 above)
Comment (ChristopherDavenport) [reply client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: Was future proofing for eventual testing I'm sure might become necessary.

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala (hunk 1 above)
Comment (ChristopherDavenport) [reply client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: The problem with methods like that, is that then users can't recreate what we've done here. But if we expose the private method then a user could get it wrong. What about if we exposed an `addCookies` which adds multiple cookies at the same time?

Comment (ChristopherDavenport) [reply client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: Added `addCookies`

**Comment by ChristopherDavenport** [comment]
Failure unrelated I believe.

File: client/src/main/scala/org/http4s/client/middleware/CookieJar.scala [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        val value = CookieValue(httpDate, expires, newCookie)
+        m + (key -> value)
+      case None => // Ignore Cookies We Can't get a domain for
+        m
+    }
+
+  private[middleware] def responseCookieToRequestCookie(r: ResponseCookie): RequestCookie =
+    RequestCookie(r.name, r.content)
+
+  private[middleware] def cookieAppliesToRequest[N[_]](
```
Comment (hamnis) [inline client/src/main/scala/org/http4s/client/middleware/CookieJar.scala]: This should also consider Secure cookies vs unsecure cookies

**Comment by ChristopherDavenport** [comment]
Question for our future selves: Now that this middleware is available should the `propagateCookies` that is present in the `FollowRedirect` be removed in favor of just using this middleware?

**Comment by rossabaker** [comment]
That came up at the time this middleware was proposed, and I don't think we ever answered it adequately.  We don't have any mechanism for one middleware implying another, at correct priority, without duplication.  Such a mechanism would solve these problems, but this is the first case that really would benefit from such, and it's a lot of machinery.

**Comment by hamnis** [comment]
> Question for our future selves: Now that this middleware is available should the `propagateCookies` that is present in the `FollowRedirect` be removed in favor of just using this middleware?

Lets not do this in 0.21.

**Pull request merged**